### PR TITLE
Fixes #643, implements __getstate__ in python API

### LINKED ISSUE #643: python bindings: InferenceSession can't be pickled
**Describe the bug**
TypeError: can't pickle onnxruntime.capi.onnxruntime_pybind11_state.InferenceSession objects

**System information**
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Debian
- ONNX Runtime installed from (source or binary): github
- ONNX Runtime version: 
- Python version: 3.7.2
- GCC/Compiler version (if compiling from source):

**To Reproduce**
pickle a InferenceSession

**Expected behavior**
Pickle by saving the bytes.

### PR DESCRIPTION
**Pull request by xadupre**
- add option to avoid cleaning initializers from the original ONNX graph
- add a function to serialize an InferenceSession into ONNX
- implements __getstate__, __setstate__

### CODE REVIEW

**Comment by yufenglee** [comment]
What's the purpose of serializing inferencesession? We can always create it from a model.

**Comment by xadupre** [comment]
Many people use pickle to save and restore objects. I did want to do it to test some converters. I then realized InferenceSession cannot return the ONNX content it was initialized with. This PR also addresses that.

**Comment by xadupre** [comment]
I'll need this PR to play with onnxruntime and dask.

**Comment by xadupre** [comment]
This PR makes it possible to retrieve the ONNX file from InferenceSession, that's how an InferenceSession is pickled. It could be done otherwise but it would mean to keep the original ONNX in memory, so the ONNX data could be duplicated 3 times in memory, almost 2 times in onnxruntime, 1 time in python. InferencSession cannot be picked if clean_initializers is true. So there is a way to disable the feature. Do you have an example in mind with databases and security? ONNX file which could be altered too. onnxruntime and databases seems independant topics to me.

File: onnxruntime/python/session.py [hunk 1]
```diff
@@ -79,3 +80,26 @@ def end_profiling(self):
         :meth:`onnxruntime.SessionOptions.enable_profiling`.
         """
         return self._sess.end_profiling()
+
+    def to_onnx(self):
```
Comment (jignparm) [inline onnxruntime/python/session.py]: if it's returning bytes[], why not call it **to_bytes()**? **to_onnx** sounds too mysterious -- unless it's doing some special processing besides returning byte representation.

**Comment by jignparm** [comment]
Resetting by vote, to remove explicit block (CI still needs to pass).

The **ability** to pickle/unpickle does not seem unsafe. The actual loading of pickled model in a production system is deemed to be unsafe.

Is reloading a model from file directly, and starting a new inference session, an option for DASK? Or is pickling required?

**Comment by nobody-cheng** [comment]
```python
class inference(object):
    def __init__(self):
        os.environ['CUDA_VISIBLE_DEVICES'] = '7'
        self.features = list()
        self.ort_sess = onnxruntime.InferenceSession(args.model_path)
        self.input_name = self.ort_sess.get_inputs()[0].name

    def run(self, img_paths):
        image = preprocess(img_paths, args.height, args.width)
        if image is not None:
            feat = self.ort_sess.run(None, {self.input_name: image})[0]
            feat = normalize(feat, axis=1)
            self.features.append(feat)


if __name__ == "__main__":
    args = get_parser().parse_args()
    img_paths = glob.glob(os.path.join(args.input, '*.jpg'))
    pool = Pool(processes=1)
    pool.map(inference().run, img_paths)
    pool.close()
    pool.join()
```
BUG
```bash
TypeError: can't pickle onnxruntime.capi.onnxruntime_pybind11_state.SessionOptions objects
```
how to fix bug,ths

**Comment by jasseur2017** [comment]
> ```python
> class inference(object):
>     def __init__(self):
>         os.environ['CUDA_VISIBLE_DEVICES'] = '7'
>         self.features = list()
>         self.ort_sess = onnxruntime.InferenceSession(args.model_path)
>         self.input_name = self.ort_sess.get_inputs()[0].name
> 
>     def run(self, img_paths):
>         image = preprocess(img_paths, args.height, args.width)
>         if image is not None:
>             feat = self.ort_sess.run(None, {self.input_name: image})[0]
>             feat = normalize(feat, axis=1)
>             self.features.append(feat)
> 
> 
> if __name__ == "__main__":
>     args = get_parser().parse_args()
>     img_paths = glob.glob(os.path.join(args.input, '*.jpg'))
>     pool = Pool(processes=1)
>     pool.map(inference().run, img_paths)
>     pool.close()
>     pool.join()
> ```
> 
> BUG
> 
> ```shell
> TypeError: can't pickle onnxruntime.capi.onnxruntime_pybind11_state.SessionOptions objects
> ```
> 
> how to fix bug,ths

Did anyone find a solution to this bug?

**Comment by aman0044** [comment]
The same bug `TypeError: can't pickle onnxruntime.capi.onnxruntime_pybind11_state.SessionOptions objects`
appears when I pass the OnnxRuntime object into the argument of [Ray](https://docs.ray.io/en/latest/actors.html) worker class's constructor.

**Comment by xadupre** [comment]
The idea is wrap onnxruntime into a class which handles the serialization. It should look like that. The main drawback is the onnx graph is stored twice in memory.

```
class WrapInferenceSession:

    def __init__(self, onnx_bytes):
        self.sess = InferenceSession(onnx_bytes.SerializeToString())
        self.onnx_bytes = onnx_bytes

    def run(self, *args):
        return self.sess.run(*args)

    def __getstate__(self):
        return {'onnx_bytes': self.onnx_bytes}

    def __setstate__(self, values):
        self.onnx_bytes = values['onnx_bytes']
        self.sess = InferenceSession(onnx_bytes.SerializeToString())         
```

**Pull request closed without merging**